### DISCUSSION TITLE
kexec into memtest86+ for RAM Wipe and reboot/poweroff - Cold Boot Attack Defense

### DISCUSSION THREAD

**Discussion by adrelanos** [memtest86plus/memtest86plus#266] [Enhancement]
[`ram-wipe`](https://www.kicksecure.com/wiki/Ram-wipe) ([source code](https://github.com/Kicksecure/ram-wipe)) ([design](https://www.kicksecure.com/wiki/Dev/RAM_Wipe)) is a software package that attempts to mitigate the threat of [cold boot attacks](https://en.wikipedia.org/wiki/Cold_boot_attack) has recently been developed (now primarily maintained by me).

`ram-wipe` is currently based on `dracut` and `sdmem`. `dracut` seems very suitable for this task. In essence, `ram-wipe` drops back to initrd ("exitrd") at shutdown when all devices are already unmounted very late in the shutdown process, uses `sdmem` to overwrite all RAM, then `kexec`'s into a new (the same) kernel and does another pass of RAM wipe (to wipe the ram of the previous kernel). However, `sdmem` seems [unmaintained upstream](https://github.com/cryptisk-grs/thc-secure-delete/issues/3#issuecomment-520754476) and [has other issues](https://www.kicksecure.com/wiki/Ram-wipe#ram-wipe_Known_Issues).

Research by [Tails](https://tails.boum.org/) seems to indicate that `memtest86+` is very suitable for wiping RAM at shutdown in principle. Much better than `sdmem`. See Tails blueprint [Erase memory: the memtest86+ way](https://gitlab.tails.boum.org/tails/blueprints/-/wikis/more_efficient_memory_wipe/memtest86plus).

What would be need to be done at the memtest86+ side:

* Support by memtest86+ to be booted using `kexec`.
* Accept kernel command line parameters such as:
  * `wiperamaction`:
   * Overwrite all RAM once with zeros.
    * The action to perform after wiping the RAM.
  * `wiperamaction="poweroff"`
  * `wiperamaction="reboot"`
  * `wiperamaction="halt"`
* Show a simple progress meter / messages for user.

Such a feature might be appreciated by both [Kicksecure](https://www.kicksecure.com) and Tails, which are derivatives of Debian Linux as well as other users who discussed, requested a wipe RAM at shutdown feature all over the internet over the years. (Note, that I cannot speak on behalf of Tails.)

Would you be interested to add a feature to memtest86+ to accommodate this use case or accept a patch?

**Comment by debrouxl** [comment]
Thanks for the clear, detailed write-up :)

To sum up: I'm not philosophically opposed, and there's definitely room for improving the functionality of both projects together through the attempt to fulfill this use case; however, I'll defer to Sam and Martin.

My two cents as a lesser maintainer:
* rewriting the contents of the RAM to mitigate cold boot attacks is a reasonable use case. That, and kexec boot, might even have neighbouring use cases in shared, virtualized hosting support, though one should not use shared hosting for dealing with confidential information in the first place;
* the memtest86+ passes which rewrite the entire memory range, i.e. most passes, are suitable for fulfilling it;
* a pure upstream memtest86+ implementation of such functionality would involve implementing at least:
  1) support for kexec boot, as you wrote;
  2) CLI support and backing code for selecting tests ( #58 ) + a maximum number of passes (part of #207 );
  3) the memory wipe pass itself, e.g. by a check-less, wait-less variant of the bit fade test;
  4) the action after the memory wipe pass is done (we currently only support manual reboot);
  5) add shortcuts for the init code, as part of the hardware detection and initialization, as well as the benchmarking code, is superfluous in your use case;
  6) add a special UI mode.

Maybe i. to iv. ought to be implemented upstream (I don't see them as being a significant cause of complexity or maintenance costs for upstream memtest86+, unless the kexec boot implementation is a pain and impacts the code everywhere, which is unlikely), and the rest ought to live downstream as a memtest86+ patch ?

**Comment by martinwhitaker** [comment]
I intentionally split the code base into boot+system+lib and app+tests to allow the basic run-time framework to be reused for other purposes. So maybe you could use memtest86+ as a sub-project and build your own simple application on top of that.

As memtest86+ supports all the boot protocols described in https://www.kernel.org/doc/Documentation/x86/boot.txt, there may not be anything else to do to make it work with kexec.

**Comment by x86fr** [comment]
Sounds really interesting to me. Implementing a quick configuration to do memory wipe is probably not very complex and could be implemented as a side feature, but as Martin said, if the changes requires are deeper, this probably deserves its own fork or sub-project.

What would be a reasonable delay for this memory wiping process? I guess in the seconds or tens of seconds range?

**Comment by adrelanos** [comment]
> What would be a reasonable delay for this memory wiping process? I guess in the seconds or tens of seconds range?

Yes, something like that. The quicker the better.

**Comment by x86fr** [comment]
Moved to discussion as it's too early for an issue/enhancement at this point.

**Comment by JacobCarrell** [comment]
I am not in favor of this, the entire thing assumes the attacker doesn't pull the memory while the system is operating.

From my understanding you have two options to defend against cold boot attacks:
1) Fix the ram in place in such a way that it can't be removed (IIRC, with standard memory controllers the contents of DRAM does not survive a system initialization.) 
2) Encrypt the contents of Ram (Supported in some Xeon processors)

**Reply by adrelanos** [reply]
Two scenarios.

* A) Without wiping RAM: User shuts down the system, leaves the house with notebook, gets robbed. -> Cold boot attack possible.
* B) With wiping RAM: User triggers wiping of RAM (by regular shutdown, triggered through device removal or panic key press), leaves the house with notebook, gets robbed. -> Cold boot attack not possible.

> Encrypt the contents of Ram (Supported in some Xeon processors)

Two options.

* 1) Encrypt RAM: Requires hardware support, will probably be closed source, relies on hardware to have an unbreakable implementation.
* 2) Software based: Can be implemented completely in software and Open Source.

Not an either/or choice. Can be both for defense-in-depth.

**Comment by JacobCarrell** [comment]
What have you been reading that substantiates this attack scenario?

 DRAM has to be refreshed in the tens of milliseconds range. Without cryogenic temps being involved they decay to noise in fractions of a second, and background noise within a minute. With cryogenics involved total loss is <60 seconds, and there's decay starting immediately after the first missed refresh period.

This issue should be closed, if for no other reason than it is objectively out of scope for a mem(ory)test(ing) tool.

 It's also perpetuating gross misunderstandings of DRAM*, ignores the existence of software (Secure Boot) and hardware (glue) mitigations. 

obligatory: [XKCD #538 - Security](https://xkcd.com/538/)

* assumption: this is also derived from the same misunderstanding of "I can recover a single bit, therefor all the data" that "experts" like David Guttman, who lack any understanding of grade-school probability, regularly vomit onto the keyboard and Journals.

**Reply by adrelanos** [reply]
* [Research of RAM data remanence times](https://blog.3mdeb.com/2024/2024-12-13-ram-data-decay-research/)
* [Research of RAM data remanence times, part 2](https://blog.3mdeb.com/2025/2025-01-24-ram-data-decay-research-part2/)
* [Conclusions from RAM data remanence tests](https://blog.3mdeb.com/2025/2025-02-20-conclusions-from-ram-data-remanence-tests/)

> for some specific memory modules this time may be higher; we have came across one specific module in which case it would be recommended to wait 2 minutes until the data decays to a sufficient extent,

This was tested without cryogenic.

**Reply by ArrayBolt3** [reply]
> assumption: this is also derived from the same misunderstanding of "I can recover a single bit, therefor all the data" that "experts" like David Guttman, who lack any understanding of grade-school probability, regularly vomit onto the keyboard and Journals.

I think you misunderstand the threat model of users who are interested in these kinds of features. It is true that for the average Joe, worrying about someone recovering a few bits from RAM is probably not necessary. There are plenty of people who aren't the average Joe. The probability of someone actually facing this kind of attack is entirely context-dependent, for some people like me it's near zero, for others it's not anywhere near zero.

> IIRC, with standard memory controllers the contents of DRAM does not survive a system initialization.

RAM contents remain intact during a reboot unfortunately, as shown by research done by 3mdeb.

It's also not necessary for a true cold boot attack to be used for data in RAM to still be a potential problem. Lots of people use features like suspend-to-RAM and don't power down their computers unless absolutely necessary. An attacker who makes off with a powered-on laptop can dump the RAM contents of the system (if Secure Boot is off, boot from a RAM dumping tool and you're done, if Secure Boot is on leave the system in suspend until next week's Secure Boot bypass is discovered, then boot your RAM dumping tool with the help of that bypass and you're done). The ability to quickly wipe RAM would be quite useful in combination with something like BusKill, to make sure that the machine fully powers itself off and wipes its RAM as soon as reasonably possible.

> This issue should be closed, if for no other reason than it is objectively out of scope for a mem(ory)test(ing) tool.

It's the closest thing to a dedicated well-maintained bootable RAM wiping system that exists at the moment to my awareness. It might not be what the tool was originally designed for, but it already does 95% of everything such a tool would need to do, so if upstream is amicable to the idea of the remaining 5% (or even part of it) being contributed, that reduces overall maintenance burden quite a bit and thus is worth discussing.
